src/java.base/share/classes/java/lang/invoke/AbstractValidatingLambdaMetafactory.java

```diff
@@ -366,12 +366,56 @@ private boolean isAdaptableTo(Class<?> fromType, Class<?> toType, boolean strict
                     return !strict;
                 }
             } else {
-                // both are reference types: fromType should be a superclass of toType.
-                return !strict || toType.isAssignableFrom(fromType);
+                // inline types: fromType and toType are projection types of the same inline class
+                // identity types: fromType should be a superclass of toType.
+                return !strict || canConvert(fromType, toType);
             }
         }
     }
 
+    /**
+     * Tests if {@code fromType} can be converted to {@code toType}
+     * via an identity conversion, via a widening reference conversion or
+     * via inline narrowing and widening conversions.
+     * <p>
+     * If {@code fromType} represents a class or interface, this method
+     * returns {@code true} if {@code toType} is the same as,
+     * or is a superclass or superinterface of, {@code fromType}.
+     * <p>
+     * If {@code fromType} is an inline class, this method returns {@code true}
+     * if {@code toType} is the {@linkplain Class#referenceType() reference projection type}
+     * of {@code fromType}.
+     * If {@code toType} is an inline class, this method returns {@code true}
+     * if {@code toType} is the {@linkplain Class#valueType() value projection type}
+     * of {@code fromType}.
+     * <p>
+     * Otherwise, this method returns {@code false}.
+     *
+     * @param     fromType the {@code Class} object to be converted from
+     * @param     toType the {@code Class} object to be converted to
+     * @return    {@code true} if {@code fromType} can be converted to {@code toType}
+     */
+    private boolean canConvert(Class<?> fromType, Class<?> toType) {
+        if (toType.isAssignableFrom(fromType)) {
+            return true;
+        }
+
```

Comment:
Since fromType.valueType is an Optional<Class<?>> the code could be rearranged to use is Present() and not double check for inline so many times.
If it is a performance issue it should be refactored to do the complete check of one case before the other.

Reply:
I simplified the check.  I also uncovered a bug in the spec for `Class::valueType` and `Class::referenceType` and included a spec fix in this patch.

Reply:
@RogerRiggs can you check out the delta patch:
   https://github.com/openjdk/valhalla/pull/102/commits/b295ac571e78c12e8bb80a4568aab0d79cdd3d06

Reply:
Looks good. No more comments.